parquet-hadoop/src/test/java/org/apache/parquet/statistics/TestStatistics.java

```diff
@@ -0,0 +1,419 @@
+      this.binaryGenerator = new RandomValueGenerators.RandomBinaryGenerator(RANDOM_SEED, recordCount);
+      this.fixedBinaryGenerator = new RandomValueGenerators.RandomFixedBinaryGenerator(RANDOM_SEED, recordCount, determineFixBinaryLength(super.schema));
+    }
+
+    private static int determineFixBinaryLength(MessageType schema) {
+      Type fixedBinaryField = schema.getType("fixed-binary");
+      PrimitiveType fixedBinaryFieldAsPrimitiveType = (PrimitiveType)fixedBinaryField;
+      return fixedBinaryFieldAsPrimitiveType.getTypeLength();
+    }
+
+    private static MessageType buildSchema(long seed) {
+      Random random = new Random(seed);
+      int fixedBinaryLength = random.nextInt(21) + 1;
+
+      return new MessageType("schema",
+        new PrimitiveType(OPTIONAL, INT32, "i32"),
+        new PrimitiveType(OPTIONAL, INT64, "i64"),
+        new PrimitiveType(OPTIONAL, INT96, "i96"),
+        new PrimitiveType(OPTIONAL, FLOAT, "sngl"),
+        new PrimitiveType(OPTIONAL, DOUBLE, "dbl"),
+        new PrimitiveType(OPTIONAL, BINARY, "binary"),
+        new PrimitiveType(OPTIONAL, FIXED_LEN_BYTE_ARRAY, fixedBinaryLength, "fixed-binary")
+      );
+    }
+
+    @Override
+    public void write(ParquetWriter<Group> writer) throws IOException {
+      for (int index = 0; index < recordCount; index++) {
+        Group group = new SimpleGroup(super.schema);
+
+        if (!intGenerator.isNull(index)) { group.append("i32", intGenerator.get(index)); }
+        if (!longGenerator.isNull(index)) { group.append("i64", longGenerator.get(index)); }
+        if (!int96Generator.isNull(index)) { group.append("i96", int96Generator.asBinary(index)); }
+        if (!floatGenerator.isNull(index)) { group.append("sngl", floatGenerator.get(index)); }
+        if (!doubleGenerator.isNull(index)) { group.append("dbl", doubleGenerator.get(index)); }
+        if (!binaryGenerator.isNull(index)) { group.append("binary", binaryGenerator.asReusedBinary(index)); }
+        if (!fixedBinaryGenerator.isNull(index)) { group.append("fixed-binary", fixedBinaryGenerator.asReusedBinary(index)); }
+
+        writer.write(group);
+      }
+    }
+
+    @Override
+    public void test() throws IOException {
+      Configuration configuration = new Configuration();
+      ParquetMetadata metadata = ParquetFileReader.readFooter(configuration, super.fsPath, ParquetMetadataConverter.NO_FILTER);
+      ParquetFileReader reader = new ParquetFileReader(configuration,
+        metadata.getFileMetaData(),
+        super.fsPath,
+        metadata.getBlocks(),
+        metadata.getFileMetaData().getSchema().getColumns());
+
+      PagingValidator<Integer> i32validator = new PagingValidator<Integer>(super.schema, "i32") {
+        @Override
+        public void updateStats(Statistics<Integer> statistics, ValuesReader valuesReader) {
+          statistics.updateStats(valuesReader.readInteger());
```

Comment:
I don't think the validation should call `updateStats`. Otherwise, you're delegating to the implementation you're trying to test. Instead, the test should be that the min value is less-than or equal-to each value in the page. Similarly, you should count nulls and validate that the number matches up after iterating through the page values. That way we are checking the meaning of the stats values independent of the implementation.

I think I mentioned this before, but it was probably overlooked in a sea of comments.